python/lsst/daf/butler/core/sqlRegistryDatabaseDict.py

```diff
@@ -0,0 +1,143 @@
+    key : `str`
+        The name of the field to be used as the dictionary key.  Must not be
+        present in ``value._fields``.
+    value : `type` (`namedtuple`)
+        The type used for the dictionary's values, typically a `namedtuple`.
+        Must have a ``_fields`` class attribute that is a tuple of field names
+        (i.e. as defined by `namedtuple`); these field names must also appear
+        in the ``types`` arg, and a `_make` attribute to construct it from a
+        sequence of values (again, as defined by `namedtuple`).
+    registry : `SqlRegistry`
+        A registry object with an open connection and a schema.
+    """
+
+    COLUMN_TYPES = {str: String, int: Integer, float: Float,
+                    bool: Boolean, bytes: LargeBinary, datetime: DateTime}
+
+    def __init__(self, config, types, key, value, registry):
+        self.registry = registry
+        allColumns = []
+        for name, type_ in types.items():
+            column = Column(name, self.COLUMN_TYPES.get(type_, type_), primary_key=(name == key))
+            allColumns.append(column)
+        if key in value._fields:
+            raise ValueError("DatabaseDict's key field may not be a part of the value tuple")
+        if key not in types.keys():
+            raise TypeError("No type provided for key {}".format(key))
+        if not types.keys() >= frozenset(value._fields):
+            raise TypeError("No type(s) provided for field(s) {}".format(set(value._fields) - types.keys()))
+        self._key = key
+        self._value = value
+        self._table = Table(config["table"], self.registry._schema.metadata, *allColumns)
+        self.registry._schema.metadata.create_all(self.registry._engine)
+        valueColumns = [getattr(self._table.columns, name) for name in self._value._fields]
+        keyColumn = getattr(self._table.columns, key)
+        self._getSql = select(valueColumns).where(keyColumn == bindparam("key"))
+        self._updateSql = self._table.update().where(keyColumn == bindparam("key"))
+        self._delSql = self._table.delete().where(keyColumn == bindparam("key"))
+        self._keysSql = select([keyColumn])
+        self._lenSql = select([func.count(keyColumn)])
+
+    def __getitem__(self, key):
+        with self.registry._connection.begin():
+            row = self.registry._connection.execute(self._getSql, key=key).fetchone()
+            if row is None:
+                raise KeyError("{} not found".format(key))
+            return self._value._make(row)
+
+    def __setitem__(self, key, value):
+        assert isinstance(value, self._value)
+        # Try insert first, as we expect that to be the most commmon usage
+        # pattern.
+        kwds = value._asdict()
+        kwds[self._key] = key
+        with self.registry._connection.begin():
+            try:
+                self.registry._connection.execute(self._table.insert(), **kwds)
+                return
+            except IntegrityError:
```

Comment:
I think there should be some comment here explaining why `IntegrityError` is perfectly okay to ignore. It seems scary, and you explain later on in the code why it's probably fine, but people are already wondering before they get further down.